Schema Design under New Notices docs #1196 (#1245)
nasa-gcn/gcn.nasa.gov: app/routes/docs.notices.unifiedschema.mdx

> ---

Reviewer: Rename this file to `docs.notices.schema.mdx`.

handle:
  breadcrumb: Unified Schema
---

import {
  Link,
  ProcessList,
  ProcessListHeading,
  ProcessListItem,
} from '@trussworks/react-uswds'

import { Highlight } from '~/components/Highlight'
import { Tab, Tabs } from '~/components/tabs/Tabs'

import { feature } from '~/lib/env.server'

export function loader() {
  return feature('SCHEMA') || new Response(null, { status: 404 })
}

# GCN Unified Schema

> The GCN Unified Schema provides a standardized method for organizing GCN notices with common properties across different missions following the uniform unit system. It utilizes JSON (JavaScript Object Notation), a lightweight and easily machine-readable data-interchange format, to define the notice schema. The set of core schemas serves as the foundation for creating new GCN Notices alerts, also allows the inclusion of instrument-specific parameters.

Reviewer: I don't think you've addressed this comment yet, have you? Namely, explaining the relationship between JSON (the schemaless data serialization) and JSON Schema (the standard for declaring the allowed fields and types in a JSON document).

## Fork the GitHub Repository

To begin, initiate by forking your own copy of the [nasa-gcn/gcn-schema](https://github.com/nasa-gcn/gcn-schema) repository on GitHub. Next, proceed to clone a local version and follow the project configuration instructions available [instructions](https://gcn.nasa.gov/docs/contributing).

## Setting Up Your Repository

> The repository's structure allows producers to create their own folders to define schemas on an instrument-specific level. The schema folder is named according to the format <code>gcn.notices.<i>mission</i></code>, where <i>mission</i> is your [Kafka Topic](/docs/notices/producers). Within it, you can define your own schema and example files. For example, schema definitions for IceCube alerts are located in the directory [`gcn/notices/icecube/`](https://github.com/nasa-gcn/gcn-schema/tree/main/gcn/notices/icecube).

Reviewer: Please update these instructions to point to the new GitHub primer (/docs/contributing/github).

## Schema File Naming

> Naming convention for your new schema files should follow the pattern <code><i>SchemaName</i>.schema.json</code> for schema definitions.

Reviewer: ```suggestion
Your new schema definition files should follow the naming convention <code><i>SchemaName</i>.schema.json</code>.
```

Examples corresponding to these schemas should follow the format, <code><i>SchemaName[.OptionalSpecification]</i>.example.json</code>. Both the <i>SchemaName</i> and <i>OptionalSpecification</i> example should be written in CamelCase formatting (ex., YourSchemaName, thisIsAlsoValid, etc.). This naming convention allows our [schema-browser](/docs/schema/stable/gcn) to generate pages based directly on the files in the latest tag.

## Sample Code

After setting up your local version, you can use the provided code samples to get started, which already include commonly used keys. Note that the `$schema` field in the example has the same value as the `$id` field in the schema.

> This ensures our validator script to identify which schema examples to compare. Make sure to update these files to reflect the mission and schema names. Then, you can update the `title` and `definition` fields to provide context for other users who will use this schema.

Reviewer: ```suggestion
This allows our validator script to identify which schema examples to compare. Make sure to update these files to reflect the mission and schema names. Then, you can update the `title` and `definition` fields to provide context for other users who will use this schema.
```

<Tabs>
  <Tab label="Sample.schema.json">
    <Highlight
      language="json"
      code={JSON.stringify(
        {
          $id: 'https://gcn.nasa.gov/schema/main/gcn/notices/<mission>/<schema name>.schema.json',
          $schema: 'https://json-schema.org/draft/2020-12/schema',
          type: 'object',
          title: 'Your Schema Name',
          description: 'A description for your schema',
          allOf: [
            {
              $ref: '/schema/gcn/notices/core/Alert.schema.json',
            },
          ],
          properties: {},
        },
        null,
        2
      )}
    />
  </Tab>
  <Tab label="Sample.example.json">
    <Highlight
      language="json"
      code={JSON.stringify(
        {
          $schema:
            'https://gcn.nasa.gov/schema/main/gcn/notices/<mission>/<schema name>.schema.json',
          alert_datetime: 'yyyy-mm-ddThh:mm:ssZ',
        },
        null,
        2
      )}
    />
  </Tab>
</Tabs>

## Defining Schema

> We recommend exploring our [schema-browser](/docs/schema/stable/gcn) to get an idea of what common fields have already been defined in the [Core schema](https://github.com/nasa-gcn/gcn-schema/tree/main/gcn/notices/core), as well as the [json schema documentation](https://www.learnjsonschema.com/) to familiarize yourself with the various keywords and types. A good method to start formulating your structure is to think first of your data product as JSON fields. Consider the possible values that your pipelines will have for each field. In the following example, we have a schema definition for Swift/BAT-GUANO alerts. They have included the `unevaluatedProperties` property and set it to `false`, which our validator can use to verify that there are no extra properties that are undefined in the examples:

Reviewer: ```suggestion
We recommend exploring our [schema browser](/docs/schema/stable/gcn) to get an idea of what common fields have already been defined in the [Core schema](https://github.com/nasa-gcn/gcn-schema/tree/main/gcn/notices/core), as well as the [JSON Schema documentation](https://json-schema.org/) to familiarize yourself with the various keywords and types. A good method to start formulating your structure is to think first of your data product as JSON fields. Consider the possible values that your pipelines will have for each field. In the following example, we have a schema definition for Swift/BAT-GUANO alerts. They have included the `unevaluatedProperties` property and set it to `false`, which our validator can use to verify that there are no extra properties that are undefined in the examples:
```

Also, break this long paragraph into 2-3 shorter ones.

              description:
                'Type of external trigger that launched the search, eg GW, neutrino, etc.',
            },
          },
          unevaluatedProperties: false,
        },
        null,
        2
      )}
    />
  </Tab>
  <Tab label="Alert.example.json">
    <Highlight
      language="json"
      code={JSON.stringify(
        {
          $schema:
            'https://gcn.nasa.gov/schema/main/gcn/notices/swift/bat/guano/alert.schema.json',
          mission: 'Swift',
          instrument: 'BAT-GUANO',
          messenger: 'EM',
          record_number: 1,
          alert_datetime: '2023-01-01T03:24:36.0Z',
          alert_tense: 'current',
          alert_type: 'initial',
          trigger_time: '2022-12-31T21:46:05.13Z',
          rate_snr: 15.8,
          rate_duration: 0.256,
          rate_energy_range: [15, 350],
          classification: { GRB: 1 },
          far: 0.00001,
          follow_up_event: 'Fermi 694215970',
          follow_up_type: 'GRB',
          data_archive_page:
            'https://guano.swift.psu.edu/trigger_report?id=694215995',
          id: ['694215995'],
        },
        null,
        2
      )}
    />
  </Tab>
</Tabs>

> While defining JSON schema structures, please don't hesitate to reach out with any

Reviewer: ```suggestion

While defining JSON schema structures, please don't hesitate to reach out with any
```

> questions about Schema field definitions or specific fields you are looking for.

Reviewer: Reach out to whom? Add link to the contact page.

questions about Schema field definitions or specific fields you are looking for.

## Validating Schema

> A validation pre-commit hook has been configured for this project. When you commit your updates, the validator will run to make sure that all of the schemas and their respective examples function correctly. This step also checks general formatting using the [Prettier](https://prettier.io/) formatter. To quickly format your code, run the following command in the root directory of your local repository. This will format spacing, line-breaks, etc. as needed.

Author: ```suggestion
The Git repository is configured to automatically format and validate your code at the time that you commit it. This means when you commit your updates, the validator will run to make sure that all of the schemas and their respective examples function correctly. This step also checks general formatting using the [Prettier](https://prettier.io/) formatter. To quickly format your code, run the following command in the root directory of your local repository. This will format spacing, line-breaks, etc. as needed.
```

> <Highlight language="sh" code="npx prettier -w ." />

Reviewer: They should _never_ have to run this manually.

> The Git repository is configured to automatically format and validate your code at the time that you commit it. This means when you commit your updates, the validator will run to make sure that all of the schemas and their respective examples function correctly. This step also checks general formatting using the [Prettier](https://prettier.io/) formatter. To quickly format your code, run the following command in the root directory of your local repository. This will format spacing, line-breaks, etc. as needed.

Reviewer: "To quickly format your code, run the following command in the root directory of your local repository." What command? Or do you want to just remove this sentence?

To validate your schemas and examples, execute the following command in the same location:

<Highlight language="sh" code="npm run validate" />

## Submit Your Schema

Once your schema and examples are validated, you can proceed to commit and push your local code to your GitHub repository.

> Next, create a pull request to our [GitHub repository](https://github.com/nasa-gcn/gcn-schema). The GCN team will review and provide feedback as needed. Once you have incorporated the feedback, we will merge your pull request into main branch. Congratulations on successfully creating your mission schema! It is now ready for production and send alerts to GCN as described in [New Notice Producers](/docs/notices/producers).

Author: ```suggestion
Next, create a pull request to our [GitHub repository](https://github.com/nasa-gcn/gcn-schema), for details see [GitHub primer](/docs/contributing/github). The GCN team will review and provide feedback as needed. Once you have incorporated the feedback, we will merge your pull request into main branch. Congratulations on successfully creating your mission schema! It is now ready for production and send alerts to GCN as described in [New Notice Producers](/docs/notices/producers).
```

>     Once your schema and examples are validated, you can proceed to commit and push your local code to your github repo.

Author: ```suggestion
    Once your schema and examples are validated, you can proceed to commit and push your local code to your GitHub repository.
```

>     Next, create a pull request to our [Git repository](https://github.com/nasa-gcn/gcn-schema). The GCN team will review and provide feedback as needed. Once you have incorporated the feedback, we will merge your pull request into main branch. Congratulations on successfully creating your mission schema! It is now ready for production and send alerts to GCN as described in [New Notice Producers](/docs/notices/producers).

Author: ```suggestion
    Next, create a pull request to our [GitHub repository](https://github.com/nasa-gcn/gcn-schema). The GCN team will review and provide feedback as needed. Once you have incorporated the feedback, we will merge your pull request into main branch. Congratulations on successfully creating your mission schema! It is now ready for production and send alerts to GCN as described in [New Notice Producers](/docs/notices/producers).
```

              description:
                'Type of external trigger that launched the search, eg GW, neutrino, etc.',
            },
          },
          unevaluatedProperties: false,
        },
        null,
        2
      )}
    />
  </Tab>
  <Tab label="Alert.example.json">
    <Highlight
      language="json"
      code={JSON.stringify(
        {
          $schema:
            'https://gcn.nasa.gov/schema/main/gcn/notices/swift/bat/guano/alert.schema.json',
          mission: 'Swift',
          instrument: 'BAT-GUANO',
          messenger: 'EM',
          record_number: 1,
          alert_datetime: '2023-01-01T03:24:36.0Z',
          alert_tense: 'current',
          alert_type: 'initial',
          trigger_time: '2022-12-31T21:46:05.13Z',
          rate_snr: 15.8,
          rate_duration: 0.256,
          rate_energy_range: [15, 350],
          classification: { GRB: 1 },
          far: 0.00001,
          follow_up_event: 'Fermi 694215970',
          follow_up_type: 'GRB',
          data_archive_page:
            'https://guano.swift.psu.edu/trigger_report?id=694215995',
          id: ['694215995'],
        },
        null,
        2
      )}
    />
  </Tab>
</Tabs>

> While defining JSON schema structures, please don't hesitate to reach out with any

Author: ```suggestion
While defining JSON schema structures, please don't hesitate to [reach out](https://gcn.nasa.gov/contact) with any
```

questions about Schema field definitions or specific fields you are looking for.

## Validating Schema

A validation pre-commit hook has been configured for this project. When you commit your updates, the validator will run to make sure that all of the schemas and their respective examples function correctly. This step also checks general formatting using the [Prettier](https://prettier.io/) formatter. To quickly format your code, run the following command in the root directory of your local repository. This will format spacing, line-breaks, etc. as needed.

> <Highlight language="sh" code="npx prettier -w ." />

Author: ```suggestion
```

}

# GCN Unified Schema

The GCN Unified Schema provides a standardized method for organizing GCN notices with common properties across different missions following the uniform unit system. It utilizes JSON (JavaScript Object Notation), a lightweight and easily machine-readable data-interchange format, to define the notice schema. The set of core schemas serves as the foundation for creating new GCN Notices alerts, also allows the inclusion of instrument-specific parameters.

## Getting Started

<ProcessList>
  <ProcessListItem>
    <ProcessListHeading type="h3">Clone the GitHub Repository</ProcessListHeading>
    To begin, clone your own copy of the [nasa-gcn/gcn-schema](https://github.com/nasa-gcn/gcn-schema) repository on github. Run the following commands to get your local setup ready:
    
    To clone the repository:
    <Highlight
      language="sh"
      code="git clone https://github.com/nasa-gcn/gcn-schema"
    />

    Change your working directory:
    <Highlight language="sh" code="cd gcn-schema" />

    Install the necessary packages:
    <Highlight language="sh" code="npm i" />

  </ProcessListItem>
  <ProcessListItem>

>   <ProcessListHeading type="h3">Setting-Up Your Repository</ProcessListHeading>

Reviewer: ```suggestion
  <ProcessListHeading type="h3">Setting Up Your Repository</ProcessListHeading>
```

      <Tab label="Sample.schema.json">
        <Highlight
          language="json"
          code={JSON.stringify(
            {
              $id: 'https://gcn.nasa.gov/schema/gcn/notices/<mission>/<schema name>.schema.json',
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              title: 'Your Schema Name',
              description: 'A description for your schema',
              "allOf": [
                {
                "$ref": "/schema/gcn/notices/core/Alert.schema.json"
                }
              ],
              properties: {
              }
            },
            null,
            2
          )}
        />
      </Tab>
      <Tab label="Sample.example.json">
        <Highlight
          language="json"
          code={JSON.stringify(
            {
              $schema:'https://gcn.nasa.gov/schema/gcn/notices/<mission>/<schema name>.schema.json',
              "alert_datetime":"yyyy-mm-ddThh:mm:ssZ"
            },
            null,
            2
          )}
        />
      </Tab>
    </Tabs>
</ProcessListItem>
<ProcessListItem>
  <ProcessListHeading type="h3">Defining Schema</ProcessListHeading>

>     We recommend exploring our [schema-browser](/docs/schema/stable/gcn) to get an idea of what common fields have already been defined in the [Core schema](https://github.com/nasa-gcn/gcn-schema/tree/main/gcn/notices/core), as well as the [json schema documentation](https://www.learnjsonschema.com/) to familiarize yourself with the various keywords and types. A good method to start formulating your structure is to think first of your data product as JSON fields. Consider the possible values that your pipelines will have for each field. In the following example, we have a schema definition for Swift/BAT-GUANO alerts. They have included the `unevaluatedProperties` property and set it to `false`, which our validator can use to verify that there are no extra properties that are undefined in the examples:

Author: ```suggestion
    We recommend exploring our [schema-browser](/docs/schema/stable/gcn) to get an idea of what common fields have already been defined in the [Core schema](https://github.com/nasa-gcn/gcn-schema/tree/v1.0.0/gcn/notices/core), as well as the [json schema documentation](https://www.learnjsonschema.com/) to familiarize yourself with the various keywords and types. A good method to start formulating your structure is to think first of your data product as JSON fields. Consider the possible values that your pipelines will have for each field. In the following example, we have a schema definition for Swift/BAT-GUANO alerts. They have included the `unevaluatedProperties` property and set it to `false`, which our validator can use to verify that there are no extra properties that are undefined in the examples:
```

    <Tabs>
      <Tab label="Alert.schema.json">
        <Highlight
          language="json"
          code={JSON.stringify(
            {

>               "$id": "https://gcn.nasa.gov/schema/gcn/notices/swift/bat/guano/alert.schema.json",

Author: ```suggestion
              "$id": "https://gcn.nasa.gov/schema/main/gcn/notices/swift/bat/guano/alert.schema.json",
```

            },
            null,
            2
          )}
        />
      </Tab>
      <Tab label="Alert.example.json">
        <Highlight
          language="json"
          code={JSON.stringify(
            {
              "$schema": "https://gcn.nasa.gov/schema/gcn/notices/swift/bat/guano/alert.schema.json",
              "mission": "Swift",
              "instrument": "BAT-GUANO",
              "messenger": "EM",
              "record_number": 1,
              "alert_datetime": "2023-01-01T03:24:36.0Z",
              "alert_tense": "current",
              "alert_type": "initial",
              "trigger_time": "2022-12-31T21:46:05.13Z",
              "rate_snr": 15.8,
              "rate_duration": 0.256,
              "rate_energy_range": [15, 350],
              "classification": { "GRB": 1 },
              "far": 0.00001,
              "follow_up_event": "Fermi 694215970",
              "follow_up_type": "GRB",
              "data_archive_page": "https://guano.swift.psu.edu/trigger_report?id=694215995",
              "id": ["694215995"]
            },
            null,
            2
          )}
        />
      </Tab>
    </Tabs>

>     Defining structures in JSON schema can get confusing, but we are happy to help.

Reviewer: It's good to include an offer of help here, but can we do that without saying that the process is expected to be confusing?

Author: I will re-word it!

>     Defining structures in JSON schema can get confusing, but we are happy to help.

Author: ```suggestion
While defining JSON schema structures, please don't hesitate to reach out with any questions about Schema field definitions or specific fields you are looking for.
```

  </ProcessListItem>
  <ProcessListItem>
    <ProcessListHeading type="h3">Validating Schema</ProcessListHeading>
    A validation pre-commit hook has been configured for this project. When you commit your updates, the validator will run to make sure that all of the schemas and their respective examples function correctly. This step also checks general formatting using the [Prettier](https://prettier.io/) formatter. To quickly format your code, run the following command in the root directory of your local repository. This will format spacing, line-breaks, etc. as needed.
    
    <Highlight language="sh" code="npx prettier -w ." />

    To validate your schemas and examples, execute the following command in the same location:

    <Highlight language="sh" code="npm run validate" />

  </ProcessListItem>
  <ProcessListItem>
    <ProcessListHeading type="h3">Submit Your Schema</ProcessListHeading>
    

>     Once your schema and examples are validated, you can proceed to commit and push your local code to the [Git repository](https://github.com/nasa-gcn/gcn-schema).

Reviewer: But they _can't_ push code to that repository. They can only push to a fork.
